tests/test_sqlalchemy_bulk.py

```diff
@@ -0,0 +1,185 @@
+    Base.metadata.create_all(engine)
+    return engine
+
+
+def table_insert(engine: Engine, type_: Type[Base], text: str) -> None:
+    pretty = PrettyFormat(column_parse={
+        'date': lambda text_: datetime.strptime(text_, '%Y-%m-%d').date()
+    })
+    with Session(engine) as session, session.begin():
+        session.add_all(type_(**attrs) for attrs in pretty.parse(text))
+
+
+def table_check(engine: Engine, type_: Type[Base], text: str) -> None:
+    pretty = PrettyFormat(
+        column_parse={'date': lambda text_: datetime.strptime(text_, '%Y-%m-%d').date()},
+        column_render={'date': lambda d: d.strftime('%Y-%m-%d')},
+        padding=0,
+    )
+    with Session(engine) as session, session.begin():
+        actual = MappedSimplifier().many(session.query(type_).all())
+        expected = pretty.parse(text)
+        compare(
+            actual=actual,
+            expected=expected,
+            suffix='\nShould be:\n'+pretty.render(actual, ref=expected)
+        )
+
+
+def test_insert(engine: Engine) -> None:
+    table_insert(
+        engine,
+        Weather,
+        """
+        +-------------+-------+-------+----+----------+
+        |city         |temp_lo|temp_hi|prcp|date      |
+        +-------------+-------+-------+----+----------+
+        |San Francisco|4      |5      |0.25|1994-11-27|
+        |San Francisco|43     |57     |0   |1994-11-29|
+        |Hayward      |37     |54     |None|1994-11-29|
+        +-------------+-------+-------+----+----------+
+        """
+    )
+    compare(
+        Session(engine).query(Weather).all(),
+        expected=[
+            Weather(city='San Francisco', temp_lo=4, temp_hi=5, prcp=0.25, date=date(1994, 11, 27)),
+            Weather(city='San Francisco', temp_lo=43, temp_hi=57, prcp=0, date=date(1994, 11, 29)),
+            Weather(city='Hayward', temp_lo=37, temp_hi=54, prcp=None, date=date(1994, 11, 29)),
+        ],
+        ignore_attributes=['_sa_instance_state']
+    )
+
+
+def test_render_pass(engine: Engine) -> None:
+    with Session(engine) as session, session.begin():
+        session.add_all((
+            Weather(city='San Francisco', temp_lo=4, temp_hi=5, prcp=0.25, date=date(1994, 11, 27)),
+            Weather(city='San Francisco', temp_lo=43, temp_hi=57, prcp=0, date=date(1994, 11, 29)),
+            Weather(city='Hayward', temp_lo=37, temp_hi=54, prcp=None, date=date(1994, 11, 29)),
+        ))
+    table_check(
+        engine,
+        Weather,
+        """
+        +-------------+-------+-------+----+----------+
+        |city         |temp_lo|temp_hi|prcp|date      |
+        +-------------+-------+-------+----+----------+
+        |San Francisco|4      |5      |0.25|1994-11-27|
+        |San Francisco|43     |57     |0   |1994-11-29|
+        |Hayward      |37     |54     |None|1994-11-29|
+        +-------------+-------+-------+----+----------+
+        """
+    )
+
+
+def test_render_fail(engine: Engine) -> None:
+    sample_lo_temp = 4
```

Comment:
Might be worth a comment to clarify that this variable exists specifically to test (something to do with string substitution and column widths)